src/custom/hooks/useWrapCallback.ts

```diff
@@ -0,0 +1,112 @@
+import { TransactionResponse } from '@ethersproject/providers'
+import { getChainCurrencySymbols } from '@src/custom/utils/xdai/hack'
+import { Currency, CurrencyAmount, currencyEquals, ETHER, WETH } from '@uniswap/sdk'
+import { Contract } from 'ethers'
+import { useMemo } from 'react'
+import { tryParseAmount } from 'state/swap/hooks'
+import { useTransactionAdder } from 'state/transactions/hooks'
+import { useCurrencyBalance } from 'state/wallet/hooks'
+import { useActiveWeb3React } from 'hooks/index'
+import { useWETHContract } from 'hooks/useContract'
+
+export enum WrapType {
+  NOT_APPLICABLE,
+  WRAP,
+  UNWRAP
+}
+
+interface WrapUnwrapCallback {
+  wrapType: WrapType
+  execute?: undefined | (() => Promise<void>)
+  inputError?: string
+}
+
+type TransactionAdder = ReturnType<typeof useTransactionAdder>
+
+interface GetWrapUnwrapCallback {
+  isWrap: boolean
+  balance?: CurrencyAmount
+  inputAmount?: CurrencyAmount
+  addTransaction: TransactionAdder
+  wethContract: Contract
+}
+
+const NOT_APPLICABLE = { wrapType: WrapType.NOT_APPLICABLE }
+
+function _getWrapUnwrapCallback(params: GetWrapUnwrapCallback): WrapUnwrapCallback {
+  const { native, wrapped } = getChainCurrencySymbols()
+  const { isWrap, balance, inputAmount, addTransaction, wethContract } = params
+  const symbol = isWrap ? native : wrapped
+
+  // Check if user has enough balance for wrap/unwrap
+  const sufficientBalance = (inputAmount && balance && !balance.lessThan(inputAmount)) || false
+
+  // Create wrap/unwrap callback if sufficient balance
+  let wrapUnwrapCallback: (() => Promise<void>) | undefined
+  if (sufficientBalance && inputAmount) {
+    let wrapUnwrap: () => TransactionResponse
+    let summary: string, symbol: string
+
+    if (isWrap) {
+      wrapUnwrap = () => wethContract.deposit({ value: `0x${inputAmount.raw.toString(16)}` })
+      summary = `Wrap ${inputAmount.toSignificant(6)} ${native} to ${wrapped}`
+    } else {
```

Comment:
I know this is not the goal of this PR at all, but I have a somewhat related question: is there anything preventing the user from wrapping all of their native currency?

Reply:
Yes, not related, although let me add an issue to prevent it from happening. Anyways, we should thing what to do in the that case, I think a confirm modal or something. Cause maybe there's use cases where it's fine to wrap it all (for example, u pay with other token the tx)
Also, it's not clear how much is fair to leave in the wallet.